### PR TITLE
[SYCLomatic] Add support for casting `dpct::kernel_function` to `uint64_t` (`CUfuntion`)

### PR DESCRIPTION
**Pull request by the-slow-one**
A variable of type `CUfunction` attribute can be assigned to another one of type `uint64_t`. The `CUfuntion` is migrated to `dpct::kernel_function` this will break such an assignment.

Sample CUDA code
```
CUfuntion func;
uint64_t funcAddr;
funcAddr = func;
```
nvcc will be able to compile this code. But `icpx` fails to compile the code migrated by DPCT. See below.

```
dpct::kernel_function func;
uint64_t funcAddr;
funcAdd = func; // This assignment will result in a error.
```

The reason is `CUfunction` is a pointer to a structure. And assignment of a pointer to `uint64_t` is semantically valid.

DPCT migrates `CUfunction` to `dpct::kernel_function` which is a class variable encapsulating a function pointer. Any assignment to `uint64_t` is semantically invalid.

This PR fixes this issue by adding the ability to typecast `kernel_funciton` object to `uint64_t`.

### CODE REVIEW

File: clang/test/dpct/kernel-function-typecast.cu [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+// RUN: FileCheck --match-full-lines --input-file %T/kernel-function-typecast/kernel-function-typecast.dp.cpp %s
+// RUN: %if build_lit %{icpx -c -fsycl %T/kernel-function-typecast/kernel-function-typecast.dp.cpp -o %T/kernel-function-typecast/kernel-function-typecast.dp.o %}
+
+#include <cstdint>
+#include <cuda.h>
+
+typedef uint64_t u64;
+u64 foo(CUfunction cuFunc, CUmodule cuMod) {
+  cuModuleGetFunction(&cuFunc, cuMod, "kfoo");
+  u64 function = (u64)cuFunc;
```
Comment (zhiweij1) [inline clang/test/dpct/kernel-function-typecast.cu:14]: Please add CHECK in the test.

Comment (zhiweij1) [inline clang/test/dpct/kernel-function-typecast.cu:14]: Is the typecast necessary?
```suggestion
  u64 function = cuFunc;
```

File: clang/runtime/dpct-rt/include/dpct/kernel.hpp [hunk 2]
```diff
@@ -400,6 +401,8 @@ class kernel_function {
     ptr(q, range, a, args, extra);
   }
 
+  operator uint64_t() const { return (uint64_t)this; }
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/kernel.hpp]: Do we also need a constructor which accepts uint64_t type argument?

File: clang/test/dpct/kernel-function-typecast.cu (hunk 1 above)
Comment (the-slow-one) [reply clang/test/dpct/kernel-function-typecast.cu:14]: Done

Comment (the-slow-one) [reply clang/test/dpct/kernel-function-typecast.cu:14]: ~The test is to check explicit typecasting. Now we have two line for both implicit and explicit casting.~

`nvcc` throws a compile error without explicit typecasting. Hence the explicit casting.

File: clang/test/dpct/kernel-function-typecast.cu [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+#include <cuda.h>
+
+typedef uint64_t u64;
+
+// CHECK: u64 foo(dpct::kernel_function cuFunc, dpct::kernel_library cuMod) {
+u64 foo(CUfunction cuFunc, CUmodule cuMod) {
+  // CHECK: cuFunc = dpct::get_kernel_function(cuMod, "kfoo");
+  cuModuleGetFunction(&cuFunc, cuMod, "kfoo");
+  u64 function = (u64)cuFunc;
+  u64 functionTwo = cuFunc;
```
Comment (the-slow-one) [inline clang/test/dpct/kernel-function-typecast.cu]:
```suggestion
```

File: clang/runtime/dpct-rt/include/dpct/kernel.hpp (hunk 2 above)
Comment (the-slow-one) [reply clang/runtime/dpct-rt/include/dpct/kernel.hpp]: No becasue CUfuntion is a pointer type to struct CUfunc_st and `nvcc` don't allow assigment of the following type.
```
CUfuntion func;
u64 value;
func = value;
```
Note: `nvcc` do allow the assignment other way round, like `value = func`.

**Review by zhiweij1 — APPROVED** [review]

**Review by AndyCHHuang — APPROVED** [review]
LGTM

**Review by zhimingwang36 — APPROVED** [review]
LGTM

**Pull request merged**